scipy/optimize/zeros.py

```diff
@@ -374,24 +374,24 @@ def _array_newton(func, x0, fprime, args, tol, maxiter, fprime2, full_output):
     """
     # Explicitly copy `x0` as `p` will be modified inplace, but the
     # user's array should not be altered.
-    try:
-        p = np.array(x0, copy=True, dtype=float)
-    except TypeError:
-        # can't convert complex to float
-        p = np.array(x0, copy=True)
+    p = np.array(x0, copy=True)
 
     failures = np.ones_like(p, dtype=bool)
     nz_der = np.ones_like(failures)
     if fprime is not None:
         # Newton-Raphson method
+        fval = np.asarray(func(p, *args))
+        fder = np.asarray(fprime(p, *args))
+        p = np.asarray(p, dtype=np.result_type(p, fval, fder, np.float64))
```

Comment:
I like this way of making the type big enough, but it looks like this only happens on the first iteration.  I.e. If x0, f(x0) and f'(x0) all happen to be real, but a 2nd iteration has complex values, then this won't catch it. 

Does adding it inside the loop instead solve the problem? Perhaps like the following:
```
            ...
            # Newton step
            dp = fval[nz_der] / fder[nz_der]
            if fprime2 is not None:
                fder2 = np.asarray(fprime2(p, *args))
                dp = dp / (1.0 - 0.5 * dp * fder2[nz_der] / fder[nz_der])
            # only update nonzero derivatives
            # Ensure p has a big enough type
            p = np.asarray(p, dtype=np.result_type(p, dp, np.float64))
            p[nz_der] -= dp
            ...
```

If that works, then iteration 0 wouldn't need to get special cased.

Reply:
>I like this way of making the type big enough, but it looks like this only happens on the first iteration. I.e. If x0, f(x0) and f'(x0) all happen to be real, but a 2nd iteration has complex values, then this won't catch it.

Fair enough, unnecessary special cases are generally bad practice.

>Does adding it inside the loop instead solve the problem?

This should be fine. The main issue here is `p[nz_der]` can't do type promotion on `p` so as long as we manually promote `p` before that statement, we should be fine.

Reply:
LGTM